src/Common/src/System/Collections/Generic/BidirectionalDictionary.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under item1 or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    internal sealed class BidirectionalDictionary<T1, T2> : IEnumerable<Tuple<T1, T2>>
+    {
+        private readonly Dictionary<T1, T2> _forward;
+        private readonly Dictionary<T2, T1> _backward;
+
+        public BidirectionalDictionary(int capacity)
+        {
+            _forward = new Dictionary<T1, T2>(capacity);
+            _backward = new Dictionary<T2, T1>(capacity);
+        }
+
+        public void Add(T1 item1, T2 item2)
+        {
+            Debug.Assert(!_backward.ContainsKey(item2), " No added item1 should ever have existing item2");
+            _forward.Add(item1, item2);
+            _backward.Add(item2, item1);
+        }
+
+        public bool TryGetForward(T1 item1, out T2 item2)
```

Comment:
I think it could be done with inheritance and overload resolution rules, and without having to ban the case of the types being the same. It wouldn't be worth the effort though.